rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,302 @@
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+| Added Trip | Scheduled Trip |
+| ---------- | -------------- |
+| `{"serviceDate": "2023-01-20", "id": "ADDED-123"}`|`{"serviceDate": "2023-01-20", "stop": "place-lake", "time": "01:02:30"}`|
+
+### Trip Type
+When adding a trip, there are three kinds of trips which can be added:
+- `"arrival"`: a trip arriving at `stop` at `time`
+- `"departure"`: a trip departing from `stop` at `time`
+- `"roundTrip"`: a trip departing from `stop` at `time` and expected to return to `stop`
+
+### Actor
+Each message will include an `actor` key, indicating who performed the given intervention.
+```json
+{
+  "username": "example",
+  "badgeNumber": "123456"
+}
+
+or
+
+{
+  "username": "pswartz"
+}
+```
+It is represented as an object to provide future extensibility if needed.
+
+### Operator
+Operators will be described by their badge number:
+```json
+{
+  "badgeNumber": "123"
+}
+```
+### Car
+Cars (individual parts of a train consist) will be described by their 4 digit number.
+```json
+{
+  "carNumber": "3800"
+}
+```
+### Optional T
+A value that can either be an instance of T, or `null`. For example, Optional Car can be
+```json
+{ "carNumber": "3800"}
+# or 
+null
+```
+## Events
+### Operator signed in
+At the start of their shift, operators need to confirm that they are fit-for-duty and do not have any electronic devices. They do this by either physically signing a paper trainsheet, or by typing their badge number into Glides.
+
+Event type: `com.mbta.ctd.glides.operator-signed-in.v1`
+Fields in the event:
+- `actor` (Actor): the inspector who signed in the operator
+- `operator` (Operator): the operator who signed in
+- `signedInAt` (RFC3999 timestamp): the time at which they signed in (separate from the `time` of the event)
+
+### Trip added
+This creates a new trip in the timesheet. It may or may not be a new trip relative to GTFS (see the splitting a train reference example).
+
+Event type: `com.mbta.ctd.glides.trip-added.v1`
+Fields in the event:
+- `actor` (Actor): the inspector adding the trip
+- `tripKey` (Trip Key): provides a unique identifier for the newly created trip
+- `stop` (string): the GTFS stop ID of the parent station at which the trip will be departing/arriving
+- `time` (HH:MM:SS time relative to the service day): when the added trip is expected to depart from/arrive at `stop`
+- `tripType` (TimeType): the relationship between `stop` and `time` for the new trip
+- `consist` (array of Optional Car, optional, default: empty list): the cars assigned to perform this trip. If a car is `null`, then no car is currently assigned to that position in the consist.
+- `operators` (list of Optional Operator, optional, default: empty list): the array of operators assigned to perform this trip. If an operator is `null`, then no operator is assigned to the car in that position in the consist.
+### Trip dropped
```

Comment:
Dropping/restoring a trip could be thought of as updating a `dropped: boolean` field in the more generic `trip-updated` event. There's a discussion elsewhere about one big update vs many small events. Dropping a trip does feel like a more significant event, but for the data I don't think it's special.

Reply:
I don't like including that: what would it meant to have a `dropped: true` along with other values in the update event? But I will include it in the "alternatives considered" section.

Reply:
Right now dropping a trip and editing the fields on a trip are separate in the UI, but they don't have to be. And it is meaningful to have data on dropped trips.

Example 1: comments are always useful, and you might want to set a comment at the same time as a dropped boolean + reason. (The UI doesn't allow that now, but the events type shouldn't prohibit it.)

Example 2, where the saved data matters even though the trip is dropped: An inspector substitutes an operator for an upcoming trip. Then they find out they don't have enough cars to do the trip, so they drop the trip. Then they find out the do infact have enough cars, so they undrop the trip, and want to retain the substitution from earlier without redoing it. (Though this example doesn't imply the dropped and operator fields would have to be in the same event.)

Even if it was part of the same event, in most cases, you'd still just send a `trip-updated` with `{dropped: true, reason: "reason"} and no other fields.

Reply:
I'll include `reason` in the other events as well (plus I can get rid of an open question!)

Reply:
Adding `reason` to `trip-added` and `trip-updated` doesn't address the question here about whether `dropped` should be a field in `trip-updated` vs its own event. My example about having `reason` in the `trip-updated` event was to show that if the events were combined, then most dropped trips would not update other fields. I wasn't trying to say that `trip-updated` needs its own reason field.

Having a `trip-updated` event that includes `dropped:true` and also edits to other fields probably isn't meaningful. But it's about as unmeaningful as having a separate dropped trip event and sending `trip-updated` right before or after `trip-dropped`, which we definitely shouldn't try to make impossible. Or as unmeaningful as sending `consist:["3901","3901"]`. In this case, I think the abstraction of making the schema treat `dropped` as just another piece of data about a trip would make handling the data much simpler, and only loses a small amount of meaning.

(Also, I have [a separate comment](https://github.com/mbta/technology-docs/pull/18#discussion_r1084582278) about the new `reason` field: I think including comments could be useful, but in that case they're comments, not reasons.)

Reply:
I think this is a good reason to separate the Trip Dropped event from the other edits: only the Trip Dropped has a reason. I also added `comment` to Trip Added and Trip Updated to cover that field.